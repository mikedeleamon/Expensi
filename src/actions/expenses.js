import uuid from 'uuid'
import database from '../firebase/firebase'


//ACTIONS
//add
//const expense = {description,note,amount,createdAt}
export const addExpense = (expense) => (
    {type:'ADD_EXPENSE',
    expense:expense}
)
export const startAddExpense = (expenseData = {}) =>{ 
    return (dispatch,getState) => {
        const uid = getState().auth.uid
        const {description='',note='',amount=0,createdAt=0} = expenseData
        const expense = {description,note,amount,createdAt}

       return database.ref(`users/${uid}/expenses`)
       
        .push(expense)
        .then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        }).catch((error)=>{console.log(error)})
    }
}

//remove
export const removeExpense = ({id}={}) => (
    {type:'REMOVE_EXPENSE',
        id:id
    })

export const startRemoveExpense = ({id}={}) => {
    return(dispatch,getState) => {
        const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`)
    .remove()
    .then(()=>{
        dispatch(removeExpense({id}))
    }).catch((error)=>{console.log(error)})
}
}

//edit
export const editExpense = (id ,updates) => ({
    type:'EDIT_EXPENSE',
    
        id:id,
        updates:updates
    
})

export const startEditExpense = (id, updates) => {
    return(dispatch,getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/expenses/${id}`)
        .update(updates)
        .then(()=>{
            dispatch(editExpense(id,updates))
        })
    }

}

export const setExpense = (expenses) => ({
    type:'SET_EXPENSE',
    expenses})

export const startSetExpenses = () => {
    return(dispatch,getState) =>{
        const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`)
    .once('value')
    .then((snapshot)=> {
        const expenses = []
        
        snapshot.forEach((childSnapshot)=>{
            expenses.push({
                id:childSnapshot.key,
            ...childSnapshot.val()
        })
        })
        dispatch(setExpense(expenses))
    })
    }
}

