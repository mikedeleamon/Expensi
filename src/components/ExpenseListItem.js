import React from 'react'
import moment from 'moment'
import numeral from 'numeral'
import {Link} from 'react-router-dom'


  const ExpsenseListItem = ({id, description, amount,createdAt}) => (
    
    <div>
      <div>
      <Link to={`/edit/${id}`}>
        <h2>{description}</h2>
      </Link>
      </div>
        <h3>{numeral(amount/100).format('$0,0.00')} - created at: {moment(createdAt).format('MMMM Do, YYYY')} </h3>
        
        
    </div>
  )

  export default ExpsenseListItem