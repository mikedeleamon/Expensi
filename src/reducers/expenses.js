const expenseStateDefault = []

export default  (state =expenseStateDefault, action) => {
    switch (action.type){
        case 'ADD_EXPENSE':
            //spread instead of using .concat()
            return [...state,action.expense]

        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id!==action.id)

        case 'EDIT_EXPENSE':
            return state.map((expense)=>{
                if (expense.id === action.id){
                    return{
                        ...expense,
                        //this will overide expenses
                        ...action.updates
                    }
                }else{
                    return expense
                }
            })
        default:
            return state
    }

}

