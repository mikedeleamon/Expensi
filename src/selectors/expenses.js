//Get visible expenses
import moment from 'moment'

export default (expenses, {text,sortBy,startDate,endDate}) =>(

    expenses.filter((expense)=> {
        //console.log(expense)
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch =  startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') :true
        const endDateMatch =  endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') :true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch && endDateMatch && textMatch
    }).sort((objOne,objTwo)=>{
        if(sortBy==='date'){
            return objOne.createdAt < objTwo.createdAt ? 1 : -1
        } if(sortBy === 'amount'){
            return objOne.amount < objTwo.amount ? 1 : -1
        }
    })
)