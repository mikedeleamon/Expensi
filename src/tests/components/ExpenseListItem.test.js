import React from "react";
import {shallow} from 'enzyme'
import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'


test('should render single item', ()=>{
   const expense = expenses[1]
    const wrapper = shallow(<ExpenseListItem key={expense.id} {...expense}/>)
    expect(wrapper).toMatchSnapshot()
})