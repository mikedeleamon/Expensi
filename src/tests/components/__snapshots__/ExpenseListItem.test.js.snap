// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render single item 1`] = `
<div>
  <div>
    <Link
      replace={false}
      to="/edit/2"
    >
      <h2>
        gas bill
      </h2>
    </Link>
  </div>
  <h3>
    $833.00
     - created at: 
    December 27th, 1969
     
  </h3>
</div>
`;
